src/Synthesiser.cpp

```diff
@@ -829,6 +829,216 @@ void Synthesiser::emitCode(std::ostream& out, const RamStatement& stmt) {
+                    break;
+            }
+
+            // Set reduction operation
+            std::string op;
+            switch (aggregate.getFunction()) {
+                case AggregateOp::MIN:
+                case AggregateOp::FMIN:
+                case AggregateOp::UMIN: {
+                    op = "min";
+                    break;
+                }
+
+                case AggregateOp::MAX:
+                case AggregateOp::FMAX:
+                case AggregateOp::UMAX: {
+                    op = "max";
+                    break;
+                }
+
+                case AggregateOp::MEAN:
+                case AggregateOp::FSUM:
+                case AggregateOp::USUM:
+                case AggregateOp::COUNT:
+
+                case AggregateOp::SUM: {
+                    op = "+";
+                    break;
+                }
+                default: fatal("Unhandled aggregate operation");
+            }
+            // res0 stores the aggregate result
+            std::string sharedVariable = "res0";
+
+            std::string type;
+            switch (getTypeAttributeAggregate(aggregate.getFunction())) {
+                case TypeAttribute::Signed: type = "RamSigned"; break;
+                case TypeAttribute::Unsigned: type = "RamUnsigned"; break;
+                case TypeAttribute::Float: type = "RamFloat"; break;
+
+                case TypeAttribute::Symbol:
+                case TypeAttribute::Record: type = "RamDomain"; break;
+            }
+            out << type << " res0 = " << init << ";\n";
+            if (aggregate.getFunction() == AggregateOp::MEAN) {
+                out << type << " res1 = 0;\n";
+                sharedVariable += ", res1";
+            }
+
+            // check whether there is an index to use
+            if (keys.empty()) {
+                out << "#pragma omp parallel for reduction(" << op << ":" << sharedVariable << ")\n";
+                out << "for(const auto& env" << identifier << " : "
+                    << "*" << relName << ") {\n";
+            } else {
+                const auto& patternsLower = aggregate.getRangePattern().first;
+                const auto& patternsUpper = aggregate.getRangePattern().second;
+
+                out << preamble.str();
+
+                out << "const " << tuple_type << " lower{{";
+                out << join(patternsLower.begin(), patternsLower.begin() + arity, ",", recWithDefault);
+                out << "}};\n";
+
+                out << "const " << tuple_type << " upper{{";
+                out << join(patternsUpper.begin(), patternsUpper.begin() + arity, ",", recWithDefault);
+                out << "}};\n";
+
+                out << "auto range = " << relName << "->"
+                    << "lowerUpperRange_" << keys << "(lower,upper," << ctxName << ");\n";
+
+                out << "auto part = range.partition();\n";
+                out << "#pragma omp parallel for reduction(" << op << ":" << sharedVariable << ")\n";
```

Comment:
Our usual approach is to use `PARALLEL_START`, which tells omp that a parallel section is starting then specify the parallel operation, such as `pragma omp for`, rather than combine them into one. I think the order of parallel section start and preamble can matter in some cases. Can the same pattern be kept here?
```suggestion
                out << "const " << tuple_type << " lower{{";
                out << join(patternsLower.begin(), patternsLower.begin() + arity, ",", recWithDefault);
                out << "}};\n";

                out << "const " << tuple_type << " upper{{";
                out << join(patternsUpper.begin(), patternsUpper.begin() + arity, ",", recWithDefault);
                out << "}};\n";

                out << "auto range = " << relName << "->"
                    << "lowerUpperRange_" << keys << "(lower,upper," << ctxName << ");\n";

                out << "auto part = range.partition();\n";
                out << "PARALLEL_START\n";
                out << preamble.str();
                out << "#pragma omp for reduction(" << op << ":" << sharedVariable << ")\n";

```

This is completely untested, so don't assume this change works or is a good idea. If it does, then there's now a redundant brace to match the `PARALLEL_END` at the end of the method that needs removing.

Reply:
Yeah the issue I thought was that I'm treating this entire section like it's running in parallel, when I really only want the for loop to run in parallel. For example, I remember running a program that contained a mean aggregate and each thread would literally divide the sum by the number of elements and each would make their own (different) entries into the table. For example, if the mean was 50.5 (5050/100), then I would get one entry into the table as 50.5, then the next one was 0.5050, and so on. I thought I could maybe protect this with "omp critical" or something but it just felt like I don't know why I'm trying to force this to fit the schema when I'm having to escape most of the stuff which should just be done sequentially anyways

Reply:
Your code does get a bit more complicated than our previous usage of omp.

`single` may do the job, without changing our existing approach. `critical` means the code is executed by one thread at a time. `single` means only a single thread executes it.

```cpp
#pragma omp single {
// ...
}
```

This is a little forced, but at least its consistent. (And maybe a later PR will overhaul our use of PARALLEL_START/END in general.)

Reply:
Okay I'll do that

Reply:
Okay done !!